README.md
[vt100](https://pkg.go.dev/github.com/gregoryv/vt100) - Provides vt100 codes and tools

## Quick start

    $ go get github.com/gregoryv/vt100/code
	
Use color

```go:
import "github.com/gregoryv/vt100"

func Example() {
	fg := vt100.ForegroundColors()
	bg := vt100.BackgroundColors()
	vt := vt100.Attributes()

	fmt.Println(fg.Red, bg.White, "hello", vt.Reset)
}
```


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on http://keepachangelog.com/en/1.0.0/
and this project adheres to http://semver.org/spec/v2.0.0.html.

## [unreleased]

- Add ParseCodes
- Add method AttributeCodes.ByName

## [0.3.0] 2022-11-27

- Add method ColorCodes.ByName

## [0.2.0] 2022-03-20

- Add types ColorCodes and AttributeCodes
- Add constructors for colors and attributes
- Remove constants for codes

## [0.1.0] 2022-03-20

- Add type Code, predefined colors and basic attributes


vt_test.go
package vt100

import (
	"testing"
)

func TestSplit(t *testing.T) {
	data := []struct {
		txt string
		exp int
	}{
		{"Hello world", 1},
		{"\033[30mHello", 2},
		{"\033[31mHello", 2},
		{"\033[32mHello", 2},
		{"Hello \033[33m World", 3},
		{"\033[0m Hello \033[35m World", 4},
	}

	for _, d := range data {
		res := Split(d.txt)
		if len(res) != d.exp {
			print("\033[0m")
			t.Errorf("Scan(%q) expected %v, got %v %#v", d.txt, d.exp, len(res), res)
		}

	}
}


code_test.go
package vt100

import (
	"fmt"
	"testing"
)

func TestParseCodes(t *testing.T) {
	if _, err := ParseCodes("red;blink"); err != nil {
		t.Fatal(err)
	}
	if _, err := ParseCodes("bgblue"); err != nil {
		t.Fatal(err)
	}
	if _, err := ParseCodes("bgnosuch"); err == nil {
		t.Fatal("bgnosuch should fail")
	}

	// unknown attributes always return reset
	if _, err := ParseCodes("bgred;nosuch"); err != nil {
		t.Fatal(err)
	}

}

func TestCode(t *testing.T) {
	c := Code(31)
	s := c.String()
	b := c.Bytes()
	if string(b) != s {
		t.Error("String and Bytes don't match")
	}
}

func TestForegroundColors(t *testing.T) {
	fg := ForegroundColors()
	bg := BackgroundColors()

	if fg.Black == bg.Black {
		t.Error("foreground and background codes are the same")
	}
}

func TestVisualInspectFG(t *testing.T) {
	defer func() {
		vt := Attributes()
		fmt.Println(vt.Reset)
	}()

	fg := ForegroundColors()
	t.Log("Black", fg.ByName("Black"), "Black")
	t.Log("Red", fg.ByName("Red"), "Red")
	t.Log("Green", fg.ByName("Green"), "Green")
	t.Log("Yellow", fg.ByName("Yellow"), "Yellow")
	t.Log("Blue", fg.ByName("Blue"), "Blue")
	t.Log("Magenta", fg.ByName("Magenta"), "Magenta")
	t.Log("Cyan", fg.ByName("Cyan"), "Cyan")
	t.Log("White", fg.ByName("White"), "White")
	// t.Fail()
}

func TestVisualInspectBG(t *testing.T) {
	defer func() {
		vt := Attributes()
		t.Log(vt.Reset)
	}()

	bg := BackgroundColors()
	t.Log(bg.Black)
	t.Log(bg.Red)
	t.Log(bg.Green)
	t.Log(bg.Yellow)
	t.Log(bg.Blue)
	t.Log(bg.Magenta)
	t.Log(bg.Cyan)
	t.Log(bg.White)
	// t.Fail()
}

func BenchmarkCode(b *testing.B) {
	c := Code(31)
	for i := 0; i < b.N; i++ {
		c.Bytes()
	}
}


example_test.go
package vt100_test

import (
	"fmt"
	"os"

	"github.com/gregoryv/vt100"
)

func Example() {
	fg := vt100.ForegroundColors()
	bg := vt100.BackgroundColors()
	vt := vt100.Attributes()

	fmt.Println(fg.Red, bg.White, "hello", vt.Reset)
}

func ExampleParseCodeBytes() {
	attr, _ := vt100.ParseCodeBytes(".*:red;blink;underscore;bgyellow")
	os.Stdout.Write(attr)
	vt := vt100.Attributes()
	fmt.Println("hello", vt.Reset)
}


code.go
package vt100

import (
	"bytes"
	"fmt"
	"strings"
)

// ParseCodeBytes groups all parsed codes as one byteslice
func ParseCodeBytes(v string) ([]byte, error) {
	codes, err := ParseCodes(v)
	if err != nil {
		return nil, err
	}
	var buf bytes.Buffer
	for _, c := range codes {
		buf.Write(c.Bytes())
	}
	return buf.Bytes(), nil
}

// ParseCodes returns a list of codes from a multicode string.
// Example: red;bgwhite;blink
func ParseCodes(v string) ([]Code, error) {
	names := strings.Split(v, ";")
	res := make([]Code, 0, len(names))
	bg := BackgroundColors()
	fg := ForegroundColors()
	at := Attributes()

	for _, name := range names {
		switch {
		case strings.HasPrefix(name, "bg"):
			name = name[2:]
			if v := bg.ByName(name); v == 0 {
				return nil, fmt.Errorf("%s: invalid background color", name)
			} else {
				res = append(res, v)
			}
		case fg.ByName(name) > 0:
			res = append(res, fg.ByName(name))

		default:
			res = append(res, at.ByName(name))
		}
	}
	return res, nil
}

func ForegroundColors() ColorCodes {
	return ColorCodes{
		Black:   30,
		Red:     31,
		Green:   32,
		Yellow:  33,
		Blue:    34,
		Magenta: 35,
		Cyan:    36,
		White:   37,
	}
}

func BackgroundColors() ColorCodes {
	return ColorCodes{
		Black:   40,
		Red:     41,
		Green:   42,
		Yellow:  43,
		Blue:    44,
		Magenta: 45,
		Cyan:    46,
		White:   47,
	}
}

type ColorCodes struct {
	Black   Code
	Red     Code
	Green   Code
	Yellow  Code
	Blue    Code
	Magenta Code
	Cyan    Code
	White   Code
}

func (c *ColorCodes) ByName(v string) Code {
	m := map[string]Code{
		"black":   c.Black,
		"red":     c.Red,
		"green":   c.Green,
		"yellow":  c.Yellow,
		"blue":    c.Blue,
		"magenta": c.Magenta,
		"cyan":    c.Cyan,
		"white":   c.White,
	}
	return m[strings.ToLower(v)]
}

func Attributes() AttributeCodes {
	return AttributeCodes{
		Reset:      0,
		Bright:     1,
		Dim:        2,
		Underscore: 4,
		Blink:      5,
		Reverse:    7,
		Hidden:     8,
	}
}

type AttributeCodes struct {
	Reset      Code
	Bright     Code
	Dim        Code
	Underscore Code
	Blink      Code
	Reverse    Code
	Hidden     Code
}

func (a *AttributeCodes) ByName(v string) Code {
	m := map[string]Code{
		"reset":      a.Reset,
		"bright":     a.Bright,
		"dim":        a.Dim,
		"underscore": a.Underscore,
		"blink":      a.Blink,
		"reverse":    a.Reverse,
		"hidden":     a.Hidden,
	}
	return m[strings.ToLower(v)]
}

type Code uint

func (me Code) Bytes() []byte {
	return []byte(me.String())
}

func (me Code) String() string {
	return fmt.Sprintf("\033[%vm", uint(me))
}
